File: src/lang/std/extrude.ts
import { v4 as uuidv4 } from 'uuid'
import { InternalFn } from './stdTypes'
import {
  ExtrudeGroup,
  ExtrudeSurface,
  SketchGroup,
  Position,
  Rotation,
} from '../executor'
import { clockwiseSign } from './std'
import { generateUuidFromHashSeed } from '../../lib/uuid'

export const extrude: InternalFn = (
  { sourceRange, engineCommandManager, code },
  length: number,
  sketchVal: SketchGroup
): ExtrudeGroup => {
  const sketch = sketchVal
  const { position, rotation } = sketchVal

  const id = generateUuidFromHashSeed(
    JSON.stringify({
      code,
      sourceRange,
      date: {
        length,
        sketchVal,
      },
    })
  )

  const extrudeSurfaces: ExtrudeSurface[] = []
  const extrusionDirection = clockwiseSign(sketch.value.map((line) => line.to))
  engineCommandManager.sendModellingCommand({
    id,
    params: [
      {
        length,
        extrusionDirection: extrusionDirection,
      },
    ],
    range: sourceRange,
    command: {
      type: 'ModelingCmdReq',
      cmd: {
        Extrude: {
          target: sketch.id,
          distance: length,
          cap: true,
        }
      },
      cmd_id: id,
      file_id: uuidv4(),
    },
  })
  
  return {
    type: 'extrudeGroup',
    value: extrudeSurfaces, // TODO, this is just an empty array now, should be deleted.
    height: length,
    position,
    rotation,
    __meta: [
      {
        sourceRange,
        pathToNode: [], // TODO
      },
      {
        sourceRange: sketchVal.__meta[0].sourceRange,
        pathToNode: sketchVal.__meta[0].pathToNode,
      },
    ],
  }
}

export const getExtrudeWallTransform: InternalFn = (
  _,
  pathName: string,
  extrudeGroup: ExtrudeGroup
): {
  position: Position
  quaternion: Rotation
} => {
  const path = extrudeGroup?.value.find((path) => path.name === pathName)
  if (!path) throw new Error(`Could not find path with name ${pathName}`)
  return {
    position: path.position,
    quaternion: path.rotation,
  }
}

File: src/wasm-lib/src/std/extrude.rs
//! Functions related to extruding.

use crate::{
    errors::{KclError, KclErrorDetails},
    executor::{ExtrudeGroup, MemoryItem},
    std::Args,
};

use anyhow::Result;

/// Extrudes by a given amount.
pub fn extrude(args: &mut Args) -> Result<MemoryItem, KclError> {
    let (length, sketch_group) = args.get_number_sketch_group()?;

    let id = uuid::Uuid::new_v4();

    let cmd = kittycad::types::ModelingCmd::Extrude {
        target: sketch_group.id,
        distance: length,
        cap: true,
    };
    args.send_modeling_cmd(id, cmd)?;

    Ok(MemoryItem::ExtrudeGroup(ExtrudeGroup {
        id,
        // TODO, this is just an empty array now, should be deleted. This
        // comment was originally in the JS code.
        value: Default::default(),
        height: length,
        position: sketch_group.position,
        rotation: sketch_group.rotation,
        meta: sketch_group.meta,
    }))
}

/// Returns the extrude wall transform.
pub fn get_extrude_wall_transform(args: &mut Args) -> Result<MemoryItem, KclError> {
    let (surface_name, extrude_group) = args.get_path_name_extrude_group()?;
    let surface = extrude_group
        .get_path_by_name(&surface_name)
        .ok_or_else(|| {
            KclError::Type(KclErrorDetails {
                message: format!(
                    "Expected a surface name that exists in the given ExtrudeGroup, found `{}`",
                    surface_name
                ),
                source_ranges: vec![args.source_range],
            })
        })?;

    Ok(MemoryItem::ExtrudeTransform {
        position: surface.get_position(),
        rotation: surface.get_rotation(),
        meta: extrude_group.meta,
    })
}
